src/supervisor.tex

```diff
@@ -1643,6 +1741,99 @@ \subsection{Addressing and Memory Protection}
 The algorithm for virtual-to-physical address translation is the same as in
 Section~\ref{sv32algorithm}, except LEVELS equals 3 and PTESIZE equals 8.
 
+\subsection{Naturally Aligned Power-of-Two Translation Contiguity}
+\label{napotpages}
+
+Bit 62 of a PTE being set indicates that a PTE represents a translation that is
+part of a range of contiguous virtual-to-physical translations with the same
+values for PTE bits 5--0.  Such ranges must be of a naturally aligned
+power-of-2 (NAPOT) granularity larger than the base page size.
+
+\begin{commentary}
+  The motivation for a NAPOT PTE is that it can be cached in a TLB as one or
+  more entries representing the contiguous region as if it were a single (large)
+  page covered by a single translation.  This compaction can help relieve
+  TLB pressure in some scenarios.  The NAPOT encoding is designed to fit within
+  the pre-existing Sv39 and Sv48 address translation algorithms so as not to
+  disrupt existing implementations or designs that choose not to implement the
+  scheme.
+\end{commentary}
+
+\begin{table*}[h!]
+\begin{center}
+\begin{tabular}{|c|c||c|l|}
+\hline
+Bit 62 & $pte.ppn[i]$      & $pte.napot\_bits$ & Meaning if $i=0$         \\
```

Comment:
I don't see where napot_bits is defined.  If the intention is for the size to be encoded implicitly by analogy with NAPOT PMP ranges, as suggested by the 1 bit in the encoding, that requires all z to be set to 0.

Reply:
This table is the definition of napot_bits.

Reply:
OK, that makes sense now.  So we could have larger NAPOT sizes in future extensions by setting the marked bit to 0, but we couldn't have smaller NAPOT sizes without bit permutation.

Years ago there was a proposal that "small" large pages could be implemented microarchitecturally, by observing all PTEs in a cache line when a fetch from memory is done.  That works up to 32 kB for a 4 kB architectural page and 8 pointers per cache line.  So maybe we don't need smaller NAPOT pages.

Reply:
The intention is that the bit position could in principle shift anywhere to indicate what size NAPOT region is needed, e.g.,

| N | PPN           | NAPOT region size |
|---|---------------|-------------------|
| 0 | `y yyyy yyyy` | 4KiB              |
| 1 | `y yyyy yy10` | 16KiB             |
| 1 | `y yyyy 1000` | 64KiB             |
| 1 | `y yy10 0000` | 256KiB            |
| 1 | `y 1000 0000` | 1MiB              |

We're starting with just one option, and reserving those z bits rather than forcing them to be zero now for NAPOT encoding, in case we decide later we'd rather use those z bits for something else more valuable.

Reply:
Makes sense.